File: src/a/pages/legacy/site-old/wwwroot/static/physics/gravity.js
// Object Representation: Define objects with properties like position (x, y, z), velocity (vx, vy, vz), and potentially mass (for more complex N-body simulations).

    const object = {
        pos: { x: 0, y: 0 }, // current position
        vel: { x: 0, y: 0 }, // current velocity
        mass: 10 // optional, for calculating gravitational force between objects
    };

//    Gravity Vector: Define a constant acceleration vector representing the force of gravity. For a simple downward pull, this would typically have a positive y-component.

    const gravity = { x: 0, y: 0.1 }; // 0.1 represents the acceleration due to gravity

// Animation Loop: Use requestAnimationFrame to create a continuous loop that updates the simulation on every frame.

    function animate() {
        // Update object properties
        // ...

        requestAnimationFrame(animate);
    }
    animate();

// Updating Velocity and Position: Inside the animation loop, apply the gravity vector to the object's velocity, and then apply the updated velocity to the object's position.

    function updateObject(obj) {
        obj.vel.x += gravity.x;
        obj.vel.y += gravity.y;

        obj.pos.x += obj.vel.x;
        obj.pos.y += obj.vel.y;
    }

// Collision Detection and Bouncing (Optional): To prevent objects from falling infinitely or to simulate bouncing, implement collision detection with boundaries (e.g., the bottom of the canvas). When a collision occurs, reverse the relevant velocity component (e.g., obj.vel.y *= -1) and potentially apply a friction or damping factor to simulate energy loss.
    // Example for bouncing off the bottom edge
    if (obj.pos.y + objectHeight > canvasHeight) {
        obj.pos.y = canvasHeight - objectHeight; // Reposition to prevent sinking
        obj.vel.y *= -0.8; // Reverse and reduce velocity for a bounce
    }

// N-Body Gravity Simulation (More Advanced):
// For simulating gravitational attraction between multiple objects (like planets in a solar system), you would use Newton's Law of Universal Gravitation to calculate the force between each pair of objects and then apply the resulting acceleration.

// Gravitational force calculation (simplified)
function calculateGravitationalForce(obj1, obj2) {
    const G = 0.01; // Gravitational constant
    const dx = obj2.pos.x - obj1.pos.x;
    const dy = obj2.pos.y - obj1.pos.y;
    const distanceSq = dx * dx + dy * dy;
    const distance = Math.sqrt(distanceSq);

    if (distance === 0) return { fx: 0, fy: 0 }; // Prevent division by zero

    const forceMagnitude = (G * obj1.mass * obj2.mass) / distanceSq;
    const fx = forceMagnitude * (dx / distance);
    const fy = forceMagnitude * (dy / distance);

    return { fx, fy };
}

File: src/a/pages/legacy/site-old/Research/Physics/gravity.py
Simulating Newtonian N-body problems in Python involves calculating the gravitational interaction between multiple celestial bodies and updating their positions and velocities over time. This is a common application in astrophysics and computational physics.
Core Principles:
Newton's Law of Universal Gravitation: Each body exerts a gravitational force on every other body, proportional to their masses and inversely proportional to the square of the distance between them.
Newton's Law of Universal Gravitation: The force (and thus acceleration) on each particle is calculated by summing the gravitational attraction from all other particles. The acceleration \(\vec{a}_{i}\) on body \(i\) due to all other bodies \(j\) is given by:

where \(G\) is the gravitational constant, \(m_{j}\) is the mass of body \(j\), and \(\vec{r}\) is the position vector. A small "softening" parameter is often added to the denominator to prevent numerical issues if two particles get too close.

Equations of Motion: These forces cause accelerations, which in turn change the bodies' velocities and positions over time.
Numerical Integration: Since there's no analytical solution for the N-body problem (for N > 2), numerical methods are used to approximate the solutions by taking small time steps.
Implementation Steps in Python:
Representing Bodies: Define a class or data structure to store each body's properties (mass, position vector, velocity vector).
Calculating Forces/Accelerations:
For each body, iterate through all other bodies to calculate the individual gravitational force exerted by each.
Sum these forces to get the net force on the current body.
Divide the net force by the body's mass to get its acceleration.
A "softening parameter" can be introduced to prevent numerical instability when bodies get extremely close.
Time Integration:
Use a numerical integration scheme (e.g., Euler method, Verlet integration, Runge-Kutta methods like those available in scipy.integrate.solve_ivp) to update the positions and velocities based on the calculated accelerations and a chosen time step.
Leap-frog integration is often used for its energy-preserving properties.
Time Integration: The positions and velocities of the particles need to be updated over time using numerical methods to solve the ordinary differential equations (ODEs) derived from Newton's laws. Common integration schemes include:Euler method: Simple but less accurate.Leapfrog or Störmer-Verlet: More stable and energy-conserving for physical systems.Runge-Kutta schemes: Higher-order algorithms that provide good accuracy, often available through scipy.integrate.solve_ivp.Vectorization: To improve performance in Python, especially for a large number of bodies (\(N\)), computations should be vectorized using NumPy arrays and matrix operations. This avoids slow Python for loops by performing calculations on entire arrays at once. 
Simulation Loop:
Repeat the force calculation and time integration steps for a desired number of time steps or until a certain simulation time is reached.
Python Libraries:
NumPy: Essential for efficient array operations, particularly for vectorized calculations of forces and accelerations, significantly improving performance compared to explicit loops.
SciPy: Provides advanced numerical integration functions (e.g., scipy.integrate.solve_ivp) for more accurate and stable time integration.
Matplotlib/VPython: Useful for visualizing the simulation results (plotting trajectories, animating the system).
Basic Implementation Outline
A basic simulation can be built from scratch using NumPy: 
Define initial conditions: Set the initial positions, velocities, and masses of all the bodies as NumPy arrays.
Create a force/acceleration function: This function takes the current positions and masses and computes the net acceleration on each body using vectorized operations.
Implement a time-stepping loop: In each step, calculate the forces, update the velocities and positions based on the chosen time integration method and a small time step dt.
Visualize the results: Use Matplotlib or other libraries (like Mayavi or Pygame) to plot or animate the particles' motion. 
Libraries and Tools
For more advanced or large-scale simulations, specialized libraries are available: 
numpy and scipy: The fundamental tools for numerical computation and integration in a from-scratch approach.
nbodykit: A massively parallel Python toolkit designed for large-scale cosmological N-body simulations, optimizing performance on supercomputers.
PySCo: A fast Particle-Mesh N-body code capable of handling large simulations (e.g., 512³ particles) efficiently on a laptop.
Pymunk or PyChrono: General-purpose physics engines if collision detection and rigid body dynamics are needed, though they might be more complex than necessary for pure Newtonian gravity simulations. 
By using these methods and libraries, one can effectively simulate the complex dynamics of Newtonian N-body systems in Python.


Example (Conceptual Outline):
Python

import numpy as np
from scipy.integrate import solve_ivp

# Gravitational constant
G = 6.674e-11  # m^3 kg^-1 s^-2

class Body:
    def __init__(self, mass, position, velocity):
        self.mass = mass
        self.position = np.array(position, dtype=float)
        self.velocity = np.array(velocity, dtype=float)

def calculate_accelerations(t, y, masses, softening):
    # y is a flattened array of all positions and velocities
    num_bodies = len(masses)
    positions = y[:num_bodies*3].reshape((num_bodies, 3))
    # velocities = y[num_bodies*3:].reshape((num_bodies, 3)) # Not directly used for acceleration calculation

    accelerations = np.zeros_like(positions)

    for i in range(num_bodies):
        for j in range(num_bodies):
            if i != j:
                r_vec = positions[j] - positions[i]
                r_mag_sq = np.sum(r_vec**2) + softening**2
                r_mag = np.sqrt(r_mag_sq)
                force_direction = r_vec / r_mag
                acceleration_mag = G * masses[j] / r_mag_sq
                accelerations[i] += acceleration_mag * force_direction
    
    # Return flattened array of velocities and accelerations for solve_ivp
    return np.concatenate((y[num_bodies*3:], accelerations.flatten()))

def simulate_n_body(bodies, total_time, dt, softening=0.1):
    initial_y = np.concatenate([body.position.flatten() for body in bodies] +
                               [body.velocity.flatten() for body in bodies])
    masses = [body.mass for body in bodies]

    # Use solve_ivp for time integration
    solution = solve_ivp(
        fun=lambda t, y: calculate_accelerations(t, y, masses, softening),
        t_span=(0, total_time),
        y0=initial_y,
        method='RK45',  # Runge-Kutta 45
        dense_output=True,
        max_step=dt
    )
    return solution


    